File: src/main/java/gargoyle/ct/ui/impl/CTSoundUpdatable.java
package gargoyle.ct.ui.impl;

import gargoyle.ct.pref.CTPreferences;
import gargoyle.ct.pref.impl.prop.CTPrefProperty;
import gargoyle.ct.sound.Audio;
import gargoyle.ct.sound.AudioClip;
import gargoyle.ct.ui.CTStateUpdatable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CTSoundUpdatable extends CTStateUpdatable {
    private static final String LOC_SOUND = "/wav/cup.aiff";
    @NotNull
    private final AudioClip clip;
    @NotNull
    private final CTPrefProperty<Boolean> sound;

    public CTSoundUpdatable(CTPreferences preferences) {
        sound = preferences.sound();
        clip = Audio.load(Audio.class.getResource(LOC_SOUND));
    }

    @Override
    protected void onStateChange(@Nullable State oldState, State newState) {
        if (oldState != null && sound.get()) {
            clip.play();
        }
    }
}

File: app/src/main/kotlin/gargoyle/ct/ui/impl/CTSoundUpdatable.kt
package gargoyle.ct.ui.impl

import gargoyle.ct.preferences.CTPreferences
import gargoyle.ct.ui.CTApp
import gargoyle.ct.ui.CTStateUpdatable
import gargoyle.ct.util.sound.Audio
import gargoyle.ct.util.sound.AudioClip
import gargoyle.ct.util.util.getResource
import kotlin.reflect.KMutableProperty0

class CTSoundUpdatable(preferences: CTPreferences) : CTStateUpdatable() {
    private val clip: AudioClip
    private val sound: KMutableProperty0<Boolean>

    init {
        sound = preferences::sound
        clip = Audio.load(CTApp::class.getResource(LOC_SOUND))
    }

    override fun onStateChange(oldState: State?, newState: State?) {
        if (oldState != null && sound.get()) {
            clip.play()
        }
    }

    companion object {
        private const val LOC_SOUND = "cup.aiff"
    }
}
